### ISSUE TITLE
Add support for configuration of the MathML printers

### ISSUE THREAD

**Issue by oscargus** [sympy/sympy#15877]
Currently, it is not possible to control the output of the MathML printers in detail (which is possible for e.g. the LaTeX printer). Two such cases are suggested in https://github.com/sympy/sympy/issues/14033#issuecomment-458457916 and https://github.com/sympy/sympy/issues/15871

The purpose here would be to introduce the additional code for easily allow adding flags to the MathML printers. Could easily be combined with e.g . https://github.com/sympy/sympy/issues/14033#issuecomment-458457916 , basically adding a setting check at https://github.com/sympy/sympy/blob/37c6d80a6080e8f72f33565fea1956c3ff28dfcf/sympy/printing/mathml.py#L789

This should probably only apply to the `MathMLPresentationPrinter` although the setting should be possible to pass through the `MathMLPrinter`as well. The reason is that the `MathMLContentPrinter` is just for content, not for presentation, and, hence, most flags will not be applicable.

**Comment by Bavish2201** [comment]
Hello, I would like to work on this. Could you please guide me on this.

**Comment by smitgajjar** [comment]
Why is the mathml printer showing floating point exponent tag even though in this function, there is a condition that if the exponent is rational, then it will be shown as mroot tag. I guess, there is some issue.

https://github.com/sympy/sympy/blob/37c6d80a6080e8f72f33565fea1956c3ff28dfcf/sympy/printing/mathml.py#L788

Example: 
``
>>> from sympy import symbols
>>> from sympy import print_mathml
>>> x=symbols('x')
>>> print_mathml(x**(1/2))
<apply>
	<power/>
	<ci>x</ci>
	<cn>0.500000000000000</cn>
</apply>

>>> print_mathml(x**(2))
<apply>
	<power/>
	<ci>x</ci>
	<cn>2</cn>
</apply>
``

**Comment by smitgajjar** [comment]
Why is the mathml printer showing floating point exponent value in the power tag, even though there is an explicit condition in this function that if the exponent is rational, then it should be shown as mroot tag form.
https://github.com/sympy/sympy/blob/37c6d80a6080e8f72f33565fea1956c3ff28dfcf/sympy/printing/mathml.py#L788

Example:
`
>>> from sympy import symbols
>>> from sympy import print_mathml
>>> x=symbols('x')
>>> print_mathml(x**(1/2))
<apply>
	<power/>
	<ci>x</ci>
	<cn>0.500000000000000</cn>
</apply>
>>> print_mathml(x**(2))
<apply>
	<power/>
	<ci>x</ci>
	<cn>2</cn>
</apply>
`
Please guide me if I am wrong. Actually, I was trying to peek into the file fimathml.py to work on this issue, but got stuck here!

**Comment by oscargus** [comment]
Good to see some interest! As there are (at least) two persons interesting I make a list of things to solve including switches to add support for (taken from the LaTeX printer):

- [ ] Add `__init__` method to `MathMLPresentationPrinter` that basically executes `MathMLPrinterBase.__init__` and extract the supported settings to local class member variables. Add documentation about the settings in `MathMLPresentationPrinter`. Add a default settings member. See `LatexPrinter` for an example.
- [ ] `fold_frac_powers` : boolean, optional
        Emit ``^{p/q}`` instead of ``^{\frac{p}{q}}`` for fractional powers. The default value is ``False``
- [ ]    `fold_short_frac` : boolean, optional
        Emit ``p / q`` instead of ``\frac{p}{q}`` when the denominator is
        simple enough (at most two terms and no powers). The default value is
        ``False``.
- [ ]    `inv_trig_style` : string, optional
        How inverse trig functions should be displayed. Can be one of
        ``abbreviated``, ``full``, or ``power``. Defaults to ``abbreviated``. (Comment: Check if there is a standard way in MathML, maybe ``full`` should not be an option here?``
- [ ]    `ln_notation` : boolean, optional
        If set to ``True``, ``\ln`` is used instead of default ``\log``.
- [ ]    `long_frac_ratio` : float or None, optional
        The allowed ratio of the width of the numerator to the width of the
        denominator before the printer breaks off long fractions. If ``None``
        (the default value), long fractions are not broken up.
- [ ]    `mat_delim` : string, optional
        The delimiter to wrap around matrices. Can be one of ``[``, ``(``, or
        the empty string. Defaults to ``[``.
- [ ] `mat_symbol_format` : string, optional
      The format used for matrix symbols. Can be "bold" or "plain". Default is "plain".
- [ ]    `mul_symbol` : string or None, optional
        The symbol to use for multiplication. Can be one of ``None``, ``ldot``,
        ``dot``, or ``times``. (Comment: check which options are feasible for MathML.)
- [ ]    `root_notation` : boolean, optional
        If set to ``False``, exponents of the form 1/n are printed in fractonal form.
        Default is ``True``, to print exponent in root form.
- [ ]    `symbol_names` : dictionary of strings mapped to symbols, optional
        Dictionary of symbols and the custom strings they should be emitted as.

For more clarification of the options, please see LatexPrinter. `mat_symbol_format` is described in #15781 . `root_notation` in https://github.com/sympy/sympy/issues/14033#issuecomment-458457916 .

https://www.w3.org/Math/ and https://www.tutorialspoint.com/mathml/ are good sources for MathML.

http://www.wiris.com/editor/demo/en/developers is a good way to see how expressions look in MathML (create expression in editor, generate MathML)
https://www.tutorialspoint.com/online_mathml_editor.php is slightly better in rendering MathML (e.g. matrices look clearly better here)

I suggest that people interested in contributing assign individual items to them. Once the first bullet is done it should be rather straightforward. Feel free to add all parameters to the documentation, `__init__`-method call, and default list even though they are not yet implemented as this will reduce the potential issues with merge conflicts later.

**Comment by smitgajjar** [comment]
I am not able to figure out the problem...Can you give the reason of not getting the required output for the shown example?

**Comment by oscargus** [comment]
Sorry, I have no idea about the inner working of these things (just happens to have some ideas from the "outside" of things). But I agree that it is surprising. It seems like the printer converts numbers to floats at some stage:

```
print_mathml(1/2, printer='presentation')
<mn>0.500000000000000</mn>


print_mathml(x/2, printer='presentation')
<mfrac>
        <mi>x</mi>
        <mn>2</mn>
</mfrac>
```

(Note that the presentation printer is the most interesting part here even though your example probably should work for the content printer as well, as 1/2 is a better representation than 0.5000000000000000000...)

**Comment by oscargus** [comment]
I think that one needs to specify it to be a Rational for the number to stay a Rational, so
```

print_mathml(x**(Rational(1,2)))
<apply>
        <root/>
        <ci>x</ci>
</apply>

print_mathml(x**(Rational(1,2)), printer='presentation')
<msqrt>
        <mi>x</mi>
</msqrt>
```

Hence, sympy does not by default keep numbers as Rational. (Not even sure if it is fully possible to bypass the Python parsing of the input, this is a thing that even the compiler may handle.)

**Comment by smitgajjar** [comment]
Ok, understood. Thanks for guiding!

**Comment by sharma-kunal** [comment]
I guess in 

```
print_mathml(1/2, printer='presentation')
```

sympy is treating 1 and 2 as math integers and thus will give answer in decimals.
While if you do S(1)/2 now 1 will be treated as a sympy object and output would be

```
print_mathml(S(1)/2, printer='presentation')
<mfrac>
	<mn>1</mn>
	<mn>2</mn>
</mfrac>

```

**Comment by sharma-kunal** [comment]
I think while doing 

```
print_mathml(x**(1/2))
```
sympy treats 1 and 2 as integers and it converts the result to 0.5
While if you change 1 to S(1), now 1 will be treated as a sympy object and thus giving the output

```
<apply>
	<root/>
	<ci>x</ci>
</apply>
```

**Comment by oscarbenjamin** [comment]
The floats here just come from the expression 1/2. This is not controllable by SymPy and is just part of basic Python. In Python 3` int/int` gives a float. Floor division is also possible with `int//int`. In SymPy we usually don't want floats and that's why as @sharma-kunal suggests we use e.g. `S(1)/2` all over the codebase. See here [https://docs.sympy.org/latest/gotchas.html#python-numbers-vs-sympy-numbers](https://docs.sympy.org/latest/gotchas.html#python-numbers-vs-sympy-numbers)

**Comment by shubhamabhang77** [comment]
I too would like to work on this issue. I am new here. So please guide me so that I can cope up with it.

**Comment by oscargus** [comment]
@shubhamabhang77 Have a look at #15901 and #15906 to get some idea. Will be easier to test once these are merged.

Then, pick one or a few flags and figure out what they mean. For example, the matrix printing code is:
https://github.com/sympy/sympy/blob/75e22267a5385591a7be80ed6cd68132b655c4e3/sympy/printing/mathml.py#L596-L607 so for `mat_delim` figure out which part generates the brackets (the line with `mfenced`), then figure out how to change that to any of the other possible bracket styles and write code that interprets the setting and generated the correct MathML. Use the resources above to see what it should be and what the generated result gives once translated to equations.

**Comment by shubhamabhang77** [comment]
will it useful to read the content from this page for this issue. https://www.w3.org/TR/MathML2/chapter3.html

**Comment by shubhamabhang77** [comment]
I have been reading on this much but I am not able to get what you said. i might be going in the wrong direction. Can you suggest any article regarding this ?

**Comment by oscargus** [comment]
Which part of it? MathML? The printer? Python? Sympy? XML?

**Comment by oscargus** [comment]
The Printer tries to find a method called _print_xxx where xxx is the class name. To see the class names, use srepr(expr). For example:

```
srepr(log(x))
Out[15]: "log(Symbol('x'))"
```
Hence, it tries to find `_print_log` and `_print_Symbol`. As `_print_log` does not exist, it tries the parent class of `log`, `Function`. Here, `_print_Function` simply prints the name of the function, line 877,
https://github.com/sympy/sympy/blob/27f2f10b965f912ea38bb189e6b498ec00efdb08/sympy/printing/mathml.py#L874-L883

Hence, to implement `ln_notation` you will need to basically copy the code from `_print_Function` to a new method `_print_log`, check for the flag and then select if you will print `ln` or `log` based on the flag. As `Function` takes an arbitrary number of arguments, it should be possible to simplify it.

A somewhat easier method in this particular case would be to add a translation for log to ln to the `translate` dictionary in `math_tag` before returning it.

(In this particular case one may want to consider the base of the log. There is an ongoing discussion to be able to specify the base in the log class and not use log(x)/log(base) as it is done today. However, at the moment it should be enough to implement it as outline above.)

**Comment by oscargus** [comment]
I'm not an expert on MathML, but I have used some "pattern matching" in my head and the editors pointed out above to get an idea what the correct output should be. Then one can always try out the result to make sure that it renders as expected.

**Comment by oscargus** [comment]
If you compare it with the generated MathML code you will be able to tell what it does.

**Comment by shubhamabhang77** [comment]
Do we have to add more attributes i.e. parameters in mathml printer as like? and I wanted to ask if I make some changes in the code of mathml.py file, what do I have to do see that  what changed in the output , basically means how can I test the code?

**Comment by oscargus** [comment]
Yes, the whole idea is to add more parameters/attributes. There are currently quite a few, although only some are implemented.

The easiest way to check the output is to execute the printer. See e.g the tests.

**Comment by shubhamabhang77** [comment]
I am not able to test the code if I make some changes ? Can you suggest something?

**Comment by smitgajjar** [comment]
In your local sympy repository, add a test case for the code to be tested
in the test file listed under bin file, then execute it

**Comment by oscargus** [comment]
Also, to simply see what comes out (to get an idea of what to add in the test), you can simply use `print_mathml` on your expression as in the third post in this thread.

(If you plan to develop directly on Github, it is in general quite tricky, and probably not feasible for this one.)

**Comment by shubhamabhang77** [comment]
Can you explain a bit more?

**Comment by oscargus** [comment]
I think it is better if you ask direct questions. Try to execute the code, modify it, and execute the modified code, and tell me where you get stuck.

**Comment by shubhamabhang77** [comment]
ohh Thanks I tried and it worked. Where are we going to use the brackets when mat_delim flag is set another brackets ? Suppose I considered this example . where do we have to use brackets in output of matrix of print_mathml?
 `
In [1]: m = Matrix([[1,2],[3,4]])

In [2]: m
Out[2]: 
⎡1  2⎤
⎢    ⎥
⎣3  4⎦

In [3]: print_mathml(m)
<matrix>
	<matrixrow>
		<cn>1</cn>
		<cn>2</cn>
	</matrixrow>
	<matrixrow>
		<cn>3</cn>
		<cn>4</cn>
	</matrixrow>
</matrix>


`

**Comment by asmeurer** [comment]
@shubhamabhang77 use three backticks ``` for multiline code blocks. Otherwise GitHub hides anything in <>.

**Comment by shubhamabhang77** [comment]
I choosed ```mat_delim``` flag to work on .

**Comment by oscargus** [comment]
If you use the presentation printer, i.e., `print_mathml(expr, printer=presentation)` it should be more clear. By default the content printer is used and then there isn't much one can change.

**Comment by shubhamabhang77** [comment]
Is all setting done in ```printer.py```  of parameters/attributes ??

**Comment by oscargus** [comment]
@shubhamabhang77 Yes, you do not need to worry about setting the parameters. You can simply get `self._settings['mat_delim']` when needed.

Right now, `print_mathml(m, printer='presentation')` gives:
```
<mfenced>
        <mtable>
                <mtr>
                        <mtd>
                                <mn>1</mn>
                        </mtd>
                        <mtd>
                                <mn>2</mn>
                        </mtd>
                </mtr>
                <mtr>
                        <mtd>
                                <mn>3</mn>
                        </mtd>
                        <mtd>
                                <mn>4</mn>
                        </mtd>
                </mtr>
        </mtable>
</mfenced>
```
However, if you want [], the `mfenced` tag should look like `<mfenced open="[" close="]">` (which one can get from http://www.wiris.com/editor/demo/en/developers , I use the "Compatible with MathML and LaTeX" section). Hence, you will need to add an attribute to the mfenced as pointed out in https://github.com/sympy/sympy/issues/15877#issuecomment-460172184 . To add an attribute you use `brac.setAttribute('open', '[')` and so on. For the option without any bracket (empty string) you just do not add the `mfenced` tag.

Since this traditionally has defaulted to '(' one should also update the default value.

If you want to see how it renders, you can use https://www.tutorialspoint.com/online_mathml_editor.php
Make sure to put the code you want to test inside `<math xmlns="http://www.w3.org/1998/Math/MathML">` and `</math>`.

**Comment by shubhamabhang77** [comment]
Is this how you want? 
```
In [2]: m
Out[2]: 
⎡1  2⎤
⎢    ⎥
⎣3  4⎦


In [3]: print_mathml(m, printer='presentation')
<mfenced close="]" open="[">
	<mtable>
		<mtr>
			<mtd>
				<mn>1</mn>
			</mtd>
			<mtd>
				<mn>2</mn>
			</mtd>
		</mtr>
		<mtr>
			<mtd>
				<mn>3</mn>
			</mtd>
			<mtd>
				<mn>4</mn>
			</mtd>
		</mtr>
	</mtable>
</mfenced>


In [4]: print_mathml(m, printer='presentation', mat_delim='(')
<mfenced close=")" open="(">
	<mtable>
		<mtr>
			<mtd>
				<mn>1</mn>
			</mtd>
			<mtd>
				<mn>2</mn>
			</mtd>
		</mtr>
		<mtr>
			<mtd>
				<mn>3</mn>
			</mtd>
			<mtd>
				<mn>4</mn>
			</mtd>
		</mtr>
	</mtable>
</mfenced>

```

**Comment by oscargus** [comment]
Yes and no. I am not sure if the default should be `'['` (it makes sense but will break what people are used to). However, that is easy to fix later (just changing a single position). The code for '[' looks good as far as I can tell (maybe one would like to print `open` before `close`, but it may be that one cannot control it and it is not important, just slightly more intuitive).

For the `'('` case it should also be OK. However, as far as I know, there is no need to specify the `open`and `close` attributes here. It will be standard rounded brackets with or without them.

Finally, the `''` (empty string) case should look like:
```
	<mtable>
		<mtr>
			<mtd>
				<mn>1</mn>
			</mtd>
			<mtd>
				<mn>2</mn>
			</mtd>
		</mtr>
		<mtr>
			<mtd>
				<mn>3</mn>
			</mtd>
			<mtd>
				<mn>4</mn>
			</mtd>
		</mtr>
	</mtable>
```
So clearly a good start!

**Comment by shubhamabhang77** [comment]
I wasn't able to control ```close``` and ```open````. Even if I change the sequence of them They were same in the output.
Is this fine for ```mat_delim``` ??
```In [3]: print_mathml(m, printer='presentation', mat_delim='[')
<mfenced close="]" open="[">
	<mtable>
		<mtr>
			<mtd>
				<mn>1</mn>
			</mtd>
			<mtd>
				<mn>2</mn>
			</mtd>
		</mtr>
		<mtr>
			<mtd>
				<mn>3</mn>
			</mtd>
			<mtd>
				<mn>4</mn>
			</mtd>
		</mtr>
	</mtable>
</mfenced>


In [4]: print_mathml(m, printer='presentation', mat_delim='(')
<mfenced>
	<mtable>
		<mtr>
			<mtd>
				<mn>1</mn>
			</mtd>
			<mtd>
				<mn>2</mn>
			</mtd>
		</mtr>
		<mtr>
			<mtd>
				<mn>3</mn>
			</mtd>
			<mtd>
				<mn>4</mn>
			</mtd>
		</mtr>
	</mtable>
</mfenced>


In [5]: print_mathml(m, printer='presentation', mat_delim='')
<mtable>
	<mtr>
		<mtd>
			<mn>1</mn>
		</mtd>
		<mtd>
			<mn>2</mn>
		</mtd>
	</mtr>
	<mtr>
		<mtd>
			<mn>3</mn>
		</mtd>
		<mtd>
			<mn>4</mn>
		</mtd>
	</mtr>
</mtable>

```

**Comment by oscargus** [comment]
I realised that maybe they were put in alphabetical order rather than the way they were entered, so that is fine.

As far as I can tell, it looks just as expected! If you prepare a pull request we can have the discussion of default value etc there. (I think the default value in the code is '[' as it is just a copy from the LaTeX printer, I also think that this makes sense and that this is a breaking change that people can live with, but let us see what other people think.

**Comment by shubhamabhang77** [comment]
the default value in the mathml is ```(``` . I saw it on https://www.w3.org/TR/MathML2/chapter3.html. In the example of section 3.3.8.3

**Comment by oscargus** [comment]
Yes, in MathML. I was thinking of the sympy code
https://github.com/sympy/sympy/blob/38174f7373a86ac92527dd7120cd92c668a38560/sympy/printing/mathml.py#L20-L34

But I think you can leave it as is and then see during the review process what the opinions are.

Please add some tests, for example those you showed above, and run the `bin/test quality` and `bin/test printing` tests at least. The rest can be run in the servers, but sometimes it is just an additional space or something so better to find that early.

**Comment by shubhamabhang77** [comment]
I have add tests in tests folder of printing, right?

**Comment by oscargus** [comment]
Yes. In `sympy/printing/tests/test_mathml.py`.

From my experience it can be better to not add it at the very end as there are currently several PRs that may add them there. Adding them somewhere else (although towards the end), may reduce the risk of having to resolve conflicts later (although they are very easy to resolve) if the other PRs are merged before.

**Comment by shubhamabhang77** [comment]
Can you tell about pull request?

**Comment by oscargus** [comment]
It doesn't look like you have cloned sympy to your repository? Maybe you only have a local copy. Either way, you will need to set git up to point to your Github account. (I always Google this as I do it quite rarely, I think there is some information here as well.)

First, you should have developed your change in a different branch of git. This will simplify things.

Once you have committed what you want you should push your branch to GitHub. Once there, you simply go to either Sympy or your local Sympy in a web browser and there will show up a button with some text asking if you want to create a pull request. It is hard to miss, just above the list of files. Press that and fill in the information requested and it should be done.

**Comment by shubhamabhang77** [comment]
Yeah right. I have just cloned the repository, not forked it.

**Comment by shubhamabhang77** [comment]
I have created a pull request. Take a look at it.

**Comment by shubhamabhang77** [comment]
For ln_notation flag, we just have to change ```log``` to ```ln``` right?

**Comment by shubhamabhang77** [comment]
Do we have use ln_notation flag for presentation printer only or for both presentation and content printer?

**Comment by oscargus** [comment]
Yes, for `ln_notation` that is correct. This should only be done for the presentation printer.

For `mul_symbol` it is about what to print as a multiplication symbol. Right now `&InvisibleTimes;` is used, i.e., just a small spacing, corresponding to the `None`option. `times` should yield `&#xD7;`, `dot`should yield `&#xB7;`, and `ldot` is a period `.`, possible with some spacing around it to distinguish it from fractional numbers. Not sure how `ldot` really should look, so see what you can come up with.

**Comment by oscargus** [comment]
And `mul_symbol` is also only presentation printer.

**Comment by shubhamabhang77** [comment]
Is this fine for ln_notation? 
```
In [1]: m = log(x)

In [2]: print_mathml(m, printer='presentation', ln_notation=False)
<mrow>
	<mi>log</mi>
	<mfenced>
		<mi>x</mi>
	</mfenced>
</mrow>


In [3]: print_mathml(m, printer='presentation', ln_notation=True)
<mrow>
	<mi>ln</mi>
	<mfenced>
		<mi>x</mi>
	</mfenced>
</mrow>

```

**Comment by shubhamabhang77** [comment]
Is this fine for ln_notation? 
```
In [1]: m = log(x)

In [2]: print_mathml(m, printer='presentation', ln_notation=False)
<mrow>
	<mi>log</mi>
	<mfenced>
		<mi>x</mi>
	</mfenced>
</mrow>


In [3]: print_mathml(m, printer='presentation', ln_notation=True)
<mrow>
	<mi>ln</mi>
	<mfenced>
		<mi>x</mi>
	</mfenced>
</mrow>
```

**Comment by shubhamabhang77** [comment]
What should I do if flag value is not valid?  for eg. ```mul_symbol='blah blah'```

**Comment by oscargus** [comment]
`ln_notation` looks fine.

I do not think that there is a single correct answer there. Either one checks it explicitly and returns an error (preferably in the initialization) or one just have a case which is not explicitly checked (typically the default case) and just fall back to that.

If you look at the LaTeX printer handling for `mul_symbol` there is a bit of code that translates the name to the actual symbol and then store the symbol in the settings under a different name as well. In this way one do not have to check every time and the code is cleaner. This can be one way to solve it (or one does something similar at the start of the method).

**Comment by shubhamabhang77** [comment]
In latex printer, If we put anything other in ```mul_symbol``` for eg. ```mul_symbol='blah blah' ``` then it's printing that same value in between two variables.

```
In [1]: z = x * y

In [2]: print(latex(z, mul_symbol='blah blah'))
xblah blahy
```

**Comment by oscargus** [comment]
That is another approach, yes. Maybe you should at least check that it is a string and throw a `TypeError` if not. In `from sympy.core.compatibility import string_types`,  `isinstance(argument, string_types)` is the way to test this that works in both Python 2.7 and 3.X.

**Comment by shubhamabhang77** [comment]
I have completed ```mul_symbol``` and ```ln_notation``` flag. Should I create a new pull request or choose next flag and then create it after all flags?

**Comment by oscargus** [comment]
It is really up to you. Probably there should be one release note entry for each anyway, so it doesn't really matter. But feel free to create one if you want to. Always good to get the code in.

**Comment by shubhamabhang77** [comment]
I have created the pull request. Take a look at it.

**Comment by shubhamabhang77** [comment]
What is the difference between ```fold_frac_powers``` and ```fold_short_frac``` ?

**Comment by oscargus** [comment]
`fold_frac_powers` is just for powers so that should be quite straightforward.

`fold_short_frac` is for any frac. I have not tried it but there seems to some logic involved when to use it, even if the flag is set. Should be possible to copy that from LaTeX or even better extract a method for determining if an expression is short enough  (maybe there already is one).

**Comment by shubhamabhang77** [comment]
What changes are expected in ```fold_frac_powers``` and in ```fold_short_frac``` ? I am not exactly getting it.

**Comment by oscargus** [comment]
With `fold_frac_powers` = True, the fraction is not written "on top" of each other but as 1/2.

In MathML, it turns out to be quite simple: `<msup><mi>x</mi><mfrac bevelled="true"><mn>1</mn><mn>2</mn></mfrac></msup>` instead of `<msup><mi>x</mi><mfrac><mn>1</mn><mn>2</mn></mfrac></msup>`. So if `fold_frac_powers`=True, set the attibute `bevelled` in the `mfrac`tag to `true`.

For `fold_short_frac` the same thing happens, not for exponents, but for a general expression, but only if the frac is deemed short enough. In the LaTeX printer the following lines does it:
https://github.com/sympy/sympy/blob/1c83dbb0164ac70ac9dc94a71fdd29e689a75dfd/sympy/printing/latex.py#L467-L477

**Comment by shubhamabhang77** [comment]
```
In [30]: q = 1 / (y)**2

In [31]: print_mathml(q, printer='presentation', fold_short_frac=False)
<msup>
	<mrow>
		<mfenced>
			<mi>y</mi>
		</mfenced>
	</mrow>
	<mn>-2</mn>
</msup>


In [32]: print_mathml(q, printer='presentation', fold_short_frac=True)
<mfrac>
	<mn>1</mn>
	<msup>
		<mrow>
			<mfenced>
				<mi>y</mi>
			</mfenced>
		</mrow>
		<mn>2</mn>
	</msup>
</mfrac>
In [33]: q = x / (y)**2

In [34]: print_mathml(q, printer='presentation', fold_short_frac=True)
<mfrac bevelled="True">
	<mi>x</mi>
	<msup>
		<mi>y</mi>
		<mn>2</mn>
	</msup>
</mfrac>


In [35]: print_mathml(q, printer='presentation', fold_short_frac=False)
<mfrac>
	<mi>x</mi>
	<msup>
		<mi>y</mi>
		<mn>2</mn>
	</msup>
</mfrac>
```
I have done this. Is this correct for ```fold_short_frac```?

**Comment by oscargus** [comment]
It is hard to say. I compared it the the LaTeX output, but there it seems like they never fold if there is a power in the denominator (see line 472).

Trying with x/y instead gives `\frac{x}{y}` for False, rendering like 
![image](https://user-images.githubusercontent.com/8114497/53180381-c4693680-35f5-11e9-9576-0dd2ceb9c8b9.png)
while True simply gives `x / y`

Either way, it seems like 32 is not correct anyway. Compare it to 31 where it is not printed like that at all, and saying that we do not want to print it like 
![image](https://user-images.githubusercontent.com/8114497/53180543-0abe9580-35f6-11e9-8fa3-b030dbf5e296.png)
gives exactly that.

34 and 35 seems OK, but if one should follow the LaTeX approach, one should not allow 34. Not sure if it is a big issue though.

Do you have any logic determining what "short" is? What is the result of `x/(y**2+x*y+2*y+x**2-y**3+1)`?

**Comment by shubhamabhang77** [comment]
```
In [1]: c = x/(y**2+x*y+2*y+x**2-y**3+1)

In [2]: c
Out[2]: 
             x              
────────────────────────────
 2          3    2          
x  + x⋅y - y  + y  + 2⋅y + 1

In [3]: print_mathml(c, printer='presentation', fold_short_frac=True)
<mfrac bevelled="True">
	<mi>x</mi>
	<mrow>
		<msup>
			<mi>x</mi>
			<mn>2</mn>
		</msup>
		<mo>+</mo>
		<mrow>
			<mi>x</mi>
			<mo>&InvisibleTimes;</mo>
			<mi>y</mi>
		</mrow>
		<mo>-</mo>
		<msup>
			<mi>y</mi>
			<mn>3</mn>
		</msup>
		<mo>+</mo>
		<msup>
			<mi>y</mi>
			<mn>2</mn>
		</msup>
		<mo>+</mo>
		<mrow>
			<mn>2</mn>
			<mo>&InvisibleTimes;</mo>
			<mi>y</mi>
		</mrow>
		<mo>+</mo>
		<mn>1</mn>
	</mrow>
</mfrac>
```
It's giving this.

For 32 given above, Will I have to add bevelled attribute or it is fully wrong?

Basically "short" means less terms in the denominator, maybe less than 3 or 4, right?

**Comment by shubhamabhang77** [comment]
```
In [3]: print_mathml(e, printer='presentation', fold_frac_powers=True)
<msup>
	<mi>x</mi>
	<mfrac bevelled="True">
		<mn>5</mn>
		<mn>3</mn>
	</mfrac>
</msup>


In [4]: print_mathml(e, printer='presentation', fold_frac_powers=False)
<msup>
	<mi>x</mi>
	<mfrac>
		<mn>5</mn>
		<mn>3</mn>
	</mfrac>
</msup>

```
What about this ```fold_frac_powers```?

**Comment by oscargus** [comment]
`fold_frac_powers` is exactly like I expects. Except that it apparently should be `"true"`, so a small t.

For 32 I think that it should either render as 31 (if you decided to exclude denominators with powers) or use bevelled. I think that it is OK to render with powers, but maybe for a numerator of 1 (as in example 31 and 32), it is still better to use the rendering in 31. For 34 and 35 it is clearly OK (as the bevelled rendering looks much better than the common LaTeX one, so it is clear where the power belongs).

I think that maybe one only would want one, at most two terms. My example expression looks like 
![image](https://user-images.githubusercontent.com/8114497/53198784-94358e00-361d-11e9-8165-713c4c7b664e.png) when rendered in MathML (again with "true"). So maybe an inital approach would be to limit it to a single Symbol (although possibly with a power etc). Not really sure how to check this though...

**Comment by shubhamabhang77** [comment]
So I should add bevelled attribute to mfrac and rest as it is in 32??
And for your example, bevelled attribute shouldn't be there because of long expression in denominator, right??  are you saying these two changes ?

**Comment by oscargus** [comment]
Yes, I think that is what I am saying. It is not clear why it doesn't happen in 32 so you may want to see what the code does there.

**Comment by shubhamabhang77** [comment]
Any more changes you want? You can give some examples to test and then we can see do we need any changes or not?

**Comment by oscargus** [comment]
Probably the best is to open a PR and we can discuss it there. I think the main thing is to make it behave similar to the LaTeX version, but I have not used the option there so I do not really know what to expect.

**Comment by shubhamabhang77** [comment]
I have made a PR. Take a look at it.

**Comment by oscargus** [comment]
With 1.4 approaching I will try to wrap this up so that it can be put in a release with introduced configurations in place.

**Comment by oscargus** [comment]
There should also be an update to the docstring describing these options.

I am also thinking about introducing a command `mathmlp` or `pmathml` that is a wrapper for `mathml(..., printer='presentation')` as that is quite tedious to write every time.

**Comment by asmeurer** [comment]
What is the status of this?

**Comment by oscargus** [comment]
I believe that there are a few parameters missing. Either one removes those from the call or one just leaves them there for order consistency with the Latex Printer. I may have time to implement a few this weekend, but most likely not all.

**Comment by asmeurer** [comment]
It's marked as blocking the release. How important is it?

**Comment by oscargus** [comment]
The problem at the moment is that it is half-finished. There are some optional arguments that are not implemented. However, they were added to keep a similar order as the other printers. Apart from that, not a big deal.

**Comment by asmeurer** [comment]
At this point only major regressions and backwards incompatibility breaks should be blockers for the release.

**Comment by oscargus** [comment]
OK! No blocker then.

**Comment by oscarbenjamin** [comment]
Unless someone wants to finish this off I think that the 1.5 milestone can be removed.

**Comment by oscarbenjamin** [comment]
Removing the 1.5 milestone

**Comment by Badrivishal** [comment]
Are there any updates required? I would like to contribute, Please guide me.

Thanking you!

**Comment by VedantGandhi04** [comment]
I would like to work on this issue. Is it still open?
